### PR TITLE
feat(pt/dpmodel): add lmdb dataloader

### LINKED ISSUE #5233: Support New Data Format: deepmd/lmdb
Add support for a new data format based on LMDB (`deepmd/lmdb`).

**Goals:**
- Prioritize LMDB implementation.
- Implement batching at the graph level (atom-count level) for efficiency.\n\n---\n*Generated by OpenClaw*

### PR DESCRIPTION
**Pull request by iProzd**
## Summary by CodeRabbit

* **New Features**
  * Full LMDB data support across training, validation and testing with PyTorch-friendly dataset/dataloader integration, per-atom-count (nloc) grouping, and deterministic distributed batching.

* **Tests**
  * Extensive unit and integration tests covering LMDB reader, dataset/dataloader, samplers, collation, and mixed-/uniform-nloc scenarios.

* **Documentation**
  * Added an example LMDB training configuration and updated test runner to handle LMDB-grouped test runs.

* **Chores**
  * Added runtime dependencies for LMDB and msgpack support.

### CODE REVIEW

File: examples/lmdb_downsample_data/input_lmdb.json [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    "start_pref_e": 0.02,
+    "limit_pref_e": 1,
+    "start_pref_f": 1000,
+    "limit_pref_f": 1,
+    "start_pref_v": 0,
+    "limit_pref_v": 0
+  },
+  "training": {
+    "training_data": {
+      "systems": "water_training.lmdb",
```
Comment (wanghan-iapcm) [inline examples/lmdb_downsample_data/input_lmdb.json:52]: the data set is not provided, user cannot run the example training.

File: source/tests/consistent/test_lmdb_data.py [hunk 2]
```diff
@@ -0,0 +1,509 @@
...
+        self.assertIn("find_energy", frame)
+        self.assertEqual(frame["coord"].dtype, np.float64)
+        self.assertEqual(frame["atype"].dtype, np.int64)
+
+    def test_uniform_nloc_single_group(self):
+        """Uniform-nloc LMDB has exactly one nloc group."""
+        reader = LmdbDataReader(self._lmdb_path, self._type_map, batch_size=2)
+        self.assertEqual(len(reader.nloc_groups), 1)
+        self.assertIn(6, reader.nloc_groups)
+        self.assertEqual(len(reader.nloc_groups[6]), 10)
```
Comment (wanghan-iapcm) [inline source/tests/consistent/test_lmdb_data.py]: these tests are not "consistency tests"

**Review by wanghan-iapcm — CHANGES_REQUESTED** [review]

**Review by wanghan-iapcm — APPROVED** [review]

File: deepmd/dpmodel/utils/lmdb_data.py [hunk 3]
```diff
@@ -0,0 +1,1527 @@
...
+    for k, v in frame.items():
+        out[_KEY_REMAP.get(k, k)] = v
+    return out
+
+
+def is_lmdb(systems: Any) -> bool:
+    """Check if systems points to an LMDB dataset."""
+    if not isinstance(systems, str):
+        return False
+    return systems.endswith(".lmdb") or Path(systems, "data.mdb").exists()
```
Comment (njzjz) [inline deepmd/dpmodel/utils/lmdb_data.py]: Checking `is_file()` is better than `exists()`.

File: deepmd/dpmodel/utils/lmdb_data.py [hunk 4]
```diff
@@ -0,0 +1,1527 @@
...
+    out = {}
+    for k, v in frame.items():
+        out[_KEY_REMAP.get(k, k)] = v
+    return out
+
+
+def is_lmdb(systems: Any) -> bool:
+    """Check if systems points to an LMDB dataset."""
+    if not isinstance(systems, str):
+        return False
```
Comment (njzjz) [inline deepmd/dpmodel/utils/lmdb_data.py]: I don't understand why this method accepts non-str inputs. What is the purpose of this design?

File: deepmd/entrypoints/test.py [hunk 5]
```diff
@@ -126,8 +148,11 @@ def test(
             systems = str((root / Path(systems)).resolve())
         else:
             systems = [str((root / Path(ss)).resolve()) for ss in systems]
-        patterns = data_params.get("rglob_patterns", None)
-        all_sys = process_systems(systems, patterns=patterns)
+        if is_lmdb(systems):
+            all_sys = [systems]
+        else:
```
Comment (njzjz) [inline deepmd/entrypoints/test.py]: It seems to me that you can move this check into `process_systems` -- it repeats for several times with the same logic.

File: deepmd/entrypoints/test.py [hunk 6]
```diff
@@ -168,61 +199,92 @@ def test(
...
-                "Global polar model is not currently supported. Please directly use the polar mode and change loss parameters."
-            )
-            err = test_polar(
-                dp, data, numb_test, detail_file, atomic=False
-            )  # YWolfeee: downward compatibility
-        log.info("# ----------------------------------------------- ")
-        err_coll.append(err)
+            data_items = [(data, system)]
+
+        for data, sys_label in data_items:
```
Comment (njzjz) [inline deepmd/entrypoints/test.py:223]: Is it possible to merge this loop into the parent loop (line 196)?

File: deepmd/pt/entrypoints/main.py [hunk 7]
```diff
@@ -146,27 +147,62 @@ def prepare_trainer_input_single(
...
+                if validation_systems
+                else None
+            )
+            train_data_single = DpLoaderSet(
+                training_systems,
+                training_dataset_params["batch_size"],
                 model_params_single["type_map"],
                 seed=rank_seed,
                 modifier=modifier,
             )
```
Comment (njzjz) [inline deepmd/pt/entrypoints/main.py]: It's better to extract the same logic into a seperated method.

**Review by njzjz — CHANGES_REQUESTED** [review]
I feel that there are too many `if...else` blocks added in this PR and they don't have a good abstract.

File: deepmd/entrypoints/test.py (hunk 6 above)
Comment (iProzd) [reply deepmd/entrypoints/test.py:223]: These two loops have different semantics — the outer loop iterates over systems, while the inner loop iterates over nloc groups within a single LMDB system (a mixed-nloc LMDB can produce multiple test groups). Merging them would require flattening all systems into a `(data, label)` list upfront, which would lose the per-system log output `("# testing system : {system}")` and make the control flow harder to follow. I think the current two-level structure is clearer, so I'd prefer to keep it as-is.

**Review by njzjz — APPROVED** [review]
